lucille-core/src/test/java/com/kmwllc/lucille/core/ZKRetryCounterTest.java

```diff
@@ -0,0 +1,145 @@
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.CuratorFrameworkFactory;
+import org.apache.curator.framework.api.BackgroundVersionable;
+import org.apache.curator.framework.api.DeleteBuilder;
+import org.apache.curator.framework.api.DeleteBuilderMain;
+import org.apache.curator.framework.recipes.shared.SharedCount;
+import org.junit.Test;
+import org.mockito.ArgumentCaptor;
+import org.mockito.MockedConstruction;
+import org.mockito.MockedStatic;
+import com.typesafe.config.ConfigFactory;
+
+public class ZKRetryCounterTest {
+
+  @Test
+  public void testRemove() throws Exception {
+    String jsonCounterString = "/LucilleCounters/bar//NON_KAFKA/runId2/jsonDoc";
+    String kafkaCounterString = "/LucilleCounters/bar/topic/runId/key___1_2";
+    try (MockedStatic<CuratorFrameworkFactory> curatorStatic = mockStatic(CuratorFrameworkFactory.class)) {
+      BackgroundVersionable backgroundVersionable = mock(BackgroundVersionable.class);
+      doThrow(Exception.class).when(backgroundVersionable).forPath(kafkaCounterString);
+      DeleteBuilderMain deleteBuilderMain = mock(DeleteBuilderMain.class);
+      when(deleteBuilderMain.deletingChildrenIfNeeded()).thenReturn(backgroundVersionable);
+      DeleteBuilder deleteBuilder = mock(DeleteBuilder.class);
+      when(deleteBuilder.quietly()).thenReturn(deleteBuilderMain);
+      CuratorFramework curatorFramework = mock(CuratorFramework.class);
+      when(curatorFramework.delete()).thenReturn(deleteBuilder);
+      curatorStatic.when(() -> CuratorFrameworkFactory.newClient(eq("foo"), any())).thenReturn(curatorFramework);
+
+      KafkaDocument kafkaDoc = spy(new KafkaDocument(Document.create("kafkaDoc"), "topic", 1, 2, "key"));
+      when(kafkaDoc.getRunId()).thenReturn("runId");
+      JsonDocument jsonDoc = new JsonDocument("jsonDoc", "runId2");
+
+      ZKRetryCounter retry3 = new ZKRetryCounter(ConfigFactory.load("ZKRetryCounterTest/config.conf"));
+      ArgumentCaptor<String> captor = ArgumentCaptor.forClass(String.class);
+
+      retry3.remove(jsonDoc);
+      retry3.remove(kafkaDoc);
+
+      verify(backgroundVersionable, times(2)).forPath(captor.capture());
+
+      assertEquals(jsonCounterString, captor.getAllValues().get(0));
+      assertEquals(kafkaCounterString, captor.getAllValues().get(1));
+    }
+  }
+
+  @Test 
+  public void testConstructor() {
+    try (MockedStatic<CuratorFrameworkFactory> curatorStatic = mockStatic(CuratorFrameworkFactory.class)) {
+      CuratorFramework curatorFramework = mock(CuratorFramework.class);
+      curatorStatic.when(() -> CuratorFrameworkFactory.newClient(eq("foo"), any())).thenReturn(curatorFramework);
+      new ZKRetryCounter(ConfigFactory.load("ZKRetryCounterTest/config.conf"));
+      verify(curatorFramework, times(1)).start();
```

Comment:
mocking lets us test things that aren't really part of the public contract of the method or constructor
in this case we're testing that the ZKRetryCounter constructor calls start on CuratorFramework but that's not really a guarantee that ZKRetryCounter aims to make. It would be reasonable for start to be called somewhere else or even for ZKRetryCounter to interact with zookeeper in a different way, not using CuratorFramework. There's a tradeoff here between improving our code coverage vs. making the tests too intrusive into implementation details that aren't part of what the method guarantees